File: js/melee-attack.js
// melee-attack.js — 近戰揮砍系統（攻擊判定綁定角色位置）
(function() {
  window.SG = window.SG || {};

  var BASE_RANGE = 160;
  var BASE_DAMAGE = 15;
  var BASE_CD = 0.6;
  var HITBOX_DURATION = 0.25;
  var ARC_ANGLE = 1.75; // ±100° 扇形（約 200° 半圓形覆蓋）

  function MeleeAttack(player) {
    this.player = player;
    this.range = BASE_RANGE;
    this.damage = BASE_DAMAGE;
    this.cd = BASE_CD;
    this.timer = 0;
    this._activeHitbox = null;
    this._activeHitbox2 = null; // 背後揮砍
    this._lastHits = [];
    this.level = 0;
  }

  MeleeAttack.prototype.update = function(dt, enemies, bosses) {
    var hits = [];
    this._lastHits = [];

    // 更新活躍的攻擊判定
    if (this._activeHitbox) {
      this._activeHitbox.timer -= dt;
      // 持續判定範圍內敵人（跟隨角色位置）
      var targets = enemies.concat(bosses);
      var angle = this.player.facingLeft ? Math.PI : 0;
      for (var i = 0; i < targets.length; i++) {
        var t = targets[i];
        if (t.hp <= 0 || this._activeHitbox.hitIds[t.id]) continue;
        var d = SG.dist(this.player, t);
        if (d > this.range + t.size / 2) continue;
        // 扇形角度檢查
        var a = Math.atan2(t.y - this.player.y, t.x - this.player.x);
        var diff = a - angle;
        while (diff > Math.PI) diff -= Math.PI * 2;
        while (diff < -Math.PI) diff += Math.PI * 2;
        if (Math.abs(diff) <= ARC_ANGLE) {
          t.hp -= this.damage;
          this._activeHitbox.hitIds[t.id] = true;
          this._lastHits.push({ x: t.x, y: t.y, dmg: this.damage });
          if (t.hp <= 0) hits.push(t);
        }
      }
      if (this._activeHitbox.timer <= 0) this._activeHitbox = null;
    }

    // 背後揮砍判定（Lv13+）
    if (this._activeHitbox2) {
      this._activeHitbox2.timer -= dt;
      var targets2 = enemies.concat(bosses);
      var backAngle = this.player.facingLeft ? 0 : Math.PI; // 反方向
      for (var i2 = 0; i2 < targets2.length; i2++) {
        var t2 = targets2[i2];
        if (t2.hp <= 0 || this._activeHitbox2.hitIds[t2.id]) continue;
        var d2 = SG.dist(this.player, t2);
        if (d2 > this.range + t2.size / 2) continue;
        var a2 = Math.atan2(t2.y - this.player.y, t2.x - this.player.x);
        var diff2 = a2 - backAngle;
        while (diff2 > Math.PI) diff2 -= Math.PI * 2;
        while (diff2 < -Math.PI) diff2 += Math.PI * 2;
        if (Math.abs(diff2) <= ARC_ANGLE) {
          t2.hp -= this.damage;
          this._activeHitbox2.hitIds[t2.id] = true;
          this._lastHits.push({ x: t2.x, y: t2.y, dmg: this.damage });
          if (t2.hp <= 0) hits.push(t2);
        }
      }
      if (this._activeHitbox2.timer <= 0) this._activeHitbox2 = null;
    }

    // 攻擊冷卻
    this.timer -= dt;
    if (this.timer <= 0 && !this._activeHitbox) {
      var targets = enemies.concat(bosses);
      if (targets.length > 0) {
        this.timer = this.cd;
        this._activeHitbox = { timer: HITBOX_DURATION, hitIds: {} };
        // Lv13+：10% 機率背後揮砍
        if (this.level >= 13 && Math.random() < 0.1) {
          this._activeHitbox2 = { timer: HITBOX_DURATION, hitIds: {} };
        }
      }
    }

    return hits;
  };

  MeleeAttack.prototype.upgradeRate = function() {
    this.cd = Math.max(0.2, this.cd - 0.05);
    this.level++;
  };

  MeleeAttack.prototype.upgradeRange = function() {
    this.range += 20;
    this.level++;
  };

  MeleeAttack.prototype.getVisual = function() {
    if (!this._activeHitbox) return null;
    var angle = this.player.facingLeft ? Math.PI : 0;
    var v = {
      x: this.player.x,
      y: this.player.y,
      angle: angle,
      range: this.range,
      progress: 1 - (this._activeHitbox.timer / HITBOX_DURATION),
      backSlash: null
    };
    if (this._activeHitbox2) {
      v.backSlash = { angle: this.player.facingLeft ? 0 : Math.PI, progress: 1 - (this._activeHitbox2.timer / HITBOX_DURATION) };
    }
    return v;
  };

  MeleeAttack.prototype.getLastHits = function() {
    var h = this._lastHits;
    this._lastHits = [];
    return h;
  };

  SG.MeleeAttack = MeleeAttack;
})();

File: unity-project/Assets/Scripts/Combat/MeleeAttack.cs
using UnityEngine;
using System.Collections.Generic;

/// <summary>
/// MeleeAttack — 騎士近戰弧形攻擊
/// 對應原始架構: melee-attack.js MeleeAttack class
/// 200° 扇形弧線攻擊，含背斬與擊退
/// </summary>
public class MeleeAttack : MonoBehaviour, IPlayerAttack
{
    [Header("Base Stats")]
    [SerializeField] private float _baseRange = 3.2f;      // 原 160px
    [SerializeField] private float _baseDamage = 15f;
    [SerializeField] private float _baseCD = 0.6f;
    [SerializeField] private float _hitboxDuration = 0.25f;
    [SerializeField] private float _arcDegrees = 200f;
    [SerializeField] private float _knockback = 2f;

    [Header("Visual")]
    [SerializeField] private GameObject _slashEffectPrefab;

    // === 狀態 ===
    private int _level = 1;
    private float _currentRange;
    private float _currentCD;
    private float _currentDamage;
    private float _timer;
    private bool _isAttacking;
    private float _hitboxTimer;
    private float _attackAngle;
    private HashSet<int> _hitEnemies = new HashSet<int>();

    // 背斬機制 (Lv13+)
    private const int BACK_SLASH_LEVEL = 13;
    private const float BACK_SLASH_CHANCE = 0.1f;

    private PlayerController _player;

    private void Start()
    {
        _currentRange = _baseRange;
        _currentCD = _baseCD;
        _currentDamage = _baseDamage;
    }

    public void Attack(PlayerController player)
    {
        _player = player;
        if (_timer > 0) return;

        _timer = _currentCD;
        _isAttacking = true;
        _hitboxTimer = _hitboxDuration;
        _hitEnemies.Clear();

        // 攻擊方向
        _attackAngle = player.FacingLeft ? 180f : 0f;

        // 背斬判定 (Lv13+)
        if (_level >= BACK_SLASH_LEVEL && Random.value < BACK_SLASH_CHANCE)
        {
            _attackAngle += 180f;
        }

        // 生成斬擊特效
        SpawnSlashEffect(player.FacingLeft);
    }

    private void Update()
    {
        if (_timer > 0) _timer -= Time.deltaTime;

        if (_isAttacking)
        {
            _hitboxTimer -= Time.deltaTime;
            CheckHits();

            if (_hitboxTimer <= 0)
            {
                _isAttacking = false;
            }
        }
    }

    /// <summary>
    /// 檢查弧形範圍內的敵人碰撞
    /// </summary>
    private void CheckHits()
    {
        var hits = Physics2D.OverlapCircleAll(transform.position, _currentRange, LayerMask.GetMask("Enemy"));

        foreach (var hit in hits)
        {
            var enemy = hit.GetComponent<EnemyBase>();
            if (enemy == null) continue;
            if (_hitEnemies.Contains(enemy.EntityID)) continue;

            // 角度判定
            Vector2 dir = (hit.transform.position - transform.position).normalized;
            float angle = Mathf.Atan2(dir.y, dir.x) * Mathf.Rad2Deg;
            float diff = Mathf.DeltaAngle(_attackAngle, angle);

            if (Mathf.Abs(diff) <= _arcDegrees / 2f)
            {
                _hitEnemies.Add(enemy.EntityID);

                float totalDamage = _currentDamage * _player.DamageMultiplier;
                enemy.TakeDamage(totalDamage);

                // 擊退
                Vector2 knockDir = dir;
                enemy.ApplyKnockback(knockDir * _knockback);

                // 傷害數字
                DamageNumberManager.Instance.Spawn(enemy.transform.position, totalDamage);
            }
        }
    }

    private void SpawnSlashEffect(bool facingLeft = true)
    {
        if (_slashEffectPrefab == null) return;

        var effect = Instantiate(_slashEffectPrefab, transform.position, Quaternion.Euler(0, 0, _attackAngle));
        var tempLocalScale = Vector3.one * (_currentRange / _baseRange);
        effect.transform.parent = transform;
        effect.transform.localScale = tempLocalScale;
        effect.transform.localPosition = facingLeft ? new Vector3(-0.5f, 0, 0) :  new Vector3(0.5f, 0, 0);
        effect.GetComponent<SpriteAnimatorController>()?.SetState(AnimState.Attack);
        effect.GetComponent<SpriteAnimatorController>()?.SetFlipY(facingLeft);
        Destroy(effect, _hitboxDuration + 0.1f);
    }

    public void Upgrade()
    {
        _level++;
        // 奇數升級 = 範圍，偶數 = 攻速
        if (_level % 2 == 1)
        {
            UpgradeRange();
        }
        else
        {
            UpgradeRate();
        }
    }

    /// <summary>
    /// 提升攻擊速度
    /// </summary>
    public void UpgradeRate()
    {
        _currentCD = Mathf.Max(0.2f, _currentCD * 0.9f);
    }

    /// <summary>
    /// 增加攻擊範圍
    /// </summary>
    public void UpgradeRange()
    {
        _currentRange += 0.3f;
    }

    public int GetLevel() => _level;

    private void OnDrawGizmosSelected()
    {
        if (!GameConfig.DEBUG_SHOW_HITBOX) return;

        Gizmos.color = new Color(0.5f, 0, 1, 0.3f);
        Gizmos.DrawWireSphere(transform.position, _currentRange);
    }
}
